backend/compact-connect/common_constructs/user_pool.py

```diff
@@ -117,19 +117,53 @@ def __init__(  # pylint: disable=too-many-arguments
 
     def add_ui_client(
         self,
-        callback_urls: list[str],
+        ui_domain_name: str,
+        environment_context: dict,
         read_attributes: ClientAttributes,
         write_attributes: ClientAttributes,
         ui_scopes: list[OAuthScope] = None,
     ):
         """
         Creates an app client for the UI to authenticate with the user pool.
 
-        :param callback_urls: The URLs that Cognito allows the UI to redirect to after authentication.
+        :param ui_domain_name: The ui domain name used to determine acceptable redirects.
+        :param environment_context: The environment context used to determine acceptable redirects.
         :param read_attributes: The attributes that the UI can read.
         :param write_attributes: The attributes that the UI can write.
         :param ui_scopes: OAuth scopes that are allowed with this client
         """
+        callback_urls = []
+        if ui_domain_name is not None:
+            callback_urls.append(f'https://{ui_domain_name}/auth/callback')
+        # This toggle will allow front-end devs to point their local UI at this environment's user pool to support
+        # authenticated actions.
+        if environment_context.get('allow_local_ui', False):
+            local_ui_port = environment_context.get('local_ui_port', '3018')
+            callback_urls.append(f'http://localhost:{local_ui_port}/auth/callback')
+        if not callback_urls:
+            raise ValueError(
+                "This app requires a callback url for its authentication path. Either provide 'domain_name' or set "
+                "'allow_local_ui' to true in this environment's context."
+            )
+
+        logout_urls = []
```

Comment:
Would `auth_urls` be a more appropriate name since this includes both Login and Logout URLs?

Reply:
Hmm, I guess `OAuthSettings` calls it `logout_urls`. I'm confused, then, why we need to include the Login URLs here?

Reply:
In the `logout_url` query param we specify where we want to be redirected to after visiting the hosted logout page. But, the `logout_url` we put there must be one of the allowed `logout_urls` listed above or else cognito will throw an error.

We are using the post logout redirect setting to chain logouts and effectively logout from both user pools if needed by visiting both hosted logout urls. 

- We need to be able to logout out of all user pools if the user clicks logout because we have no way of stopping a user from logging in to both user pools simultaneously due the hosted login pages being for separate user pools / not being aware of each other.

- We need to log out of all user pools by chaining redirects because the only way to fully logout of a cognito userpool is by visiting its logout url, so we need to rapidly visit all of them on "logout." Other methods leave the user logged in in various ways. I can elaborate more if desired.

- That all being said it is very unlikely a user will be logged in to both user pools simultaneously as it requires bookmarking one of the hosted login pages / visiting it after already logged in to one user pool, which a user would not naturally do through using the app.

- The weirdness and complexity of all of these flows comes from the potential edge case of a user logging in to both user pools simultaneously combined with cognito's opinionated logout functionality that disables us from merely logging the user out of the other user pool when they login to the new one.

Reply:
Got it. This is unintuitive enough that I think it might warrant documenting in a README. I think it might also make sense to document something on your other comment re: being logged in remotely vs locally: "Logged out from the perspective of cognito, not just locally. Local logout would be deleting the tokens. Remote logout would be making the tokens invalid / revoking them + removing http only cookies."

Thoughts?

Reply:
Ok, Ill create a readme documenting the nuances of cognito, how our two user pool set up meshes with those nuances and explaining our mitigation efforts.

Reply:
Awesome, thank you!